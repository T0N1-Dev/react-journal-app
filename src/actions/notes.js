import { addDoc, collection, doc, updateDoc, deleteDoc } from "firebase/firestore"
import { db } from "../firebase/firebaseConfig"
import { types } from "../types/types"
import { loadNotes } from "../helpers/loadNotes";
import Swal from "sweetalert2";
import { fileUpload } from "../helpers/fileUpload";

export const startNewNote = () => {
    return async ( dispatch, getState ) => {

        const uid = getState().auth.uid;
        
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime(),
        }

        Swal.fire({
            title: 'Adding...',
            text: 'Please wait...', 
            allowOutsideClick: false,
            didOpen: () => {
                Swal.showLoading();
            }
        });
        
        try {
            const docRef = await addDoc(collection(db, `${uid}/journal/notes`), newNote);
            const noteWithId = { id: docRef.id, ...newNote };
            dispatch( activeNote( docRef.id, newNote ));
            dispatch(addNoteToState(noteWithId));
        } catch (error) {
            Swal.fire('Error al añadir la nueva nota:', error, 'error');
        } finally {
            Swal.close();
        }
    }
};

export const activeNote = ( id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
});

export const addNoteToState = (note) => ({
    type: types.noteAdded,
    payload: note
});

export const startLoadingNotes = (uid) => {
    return async (dispatch) => {
        const notes = await loadNotes( uid );
        dispatch(setNotes(notes));
    }
}

export const setNotes = ( notes ) => ({
    type: types.notesLoad,
    payload: notes
});

export const startSaveNote = ( note ) => {
    return async ( dispatch, getState ) => {
        const { uid } = getState().auth;

        const noteToFirestore = { ...note };
        delete noteToFirestore.id;

        try {
            const notRef = doc(db, `${uid}/journal/notes/${note.id}`);
            await updateDoc(notRef, noteToFirestore);
            dispatch( refreshNote(note.id, noteToFirestore ));
            dispatch( activeNote(note.id, noteToFirestore));
            Swal.fire('Nota actualizada correctamente', note.title, 'success');
        } catch (error) {
            Swal.fire('Error al actualizar la nota:', error, 'error');
        }
    }
}

export const refreshNote = ( id, note ) => ({
    type: types.notesUpdated,
    payload: {
        id, 
        note: {
            id,
            ...note
        }
    }
});

export const startUploading = (file) => {
    return async (dispatch, getState) => {
        const { active: activeNote } = getState().notes;

        Swal.fire({
            title: 'Uploading...',
            text: 'Please wait...', 
            allowOutsideClick: false,
            didOpen: () => {
                Swal.showLoading();
            }
            
        });

        try {
            const fileUrl = await fileUpload( file );

            const updatedNote = {
                ...activeNote,
                url: fileUrl
            };
            dispatch( startSaveNote( updatedNote ));
        } catch (error) {
            Swal.fire('Error', 'Failed to upload the file', 'error');
        } finally {
            Swal.close();
        }
    }
}

export const startDeleting = ( id ) => {
    return (dispatch, getState) => {
         const { uid } = getState().auth;
         const notRef = doc(db, `${uid}/journal/notes/${id}`);
         
         try {
            Swal.fire({
                title: "Are you sure?",
                text: "You won't be able to revert this!",
                icon: "warning",
                showCancelButton: true,
                confirmButtonColor: "#3085d6",
                cancelButtonColor: "#d33",
                confirmButtonText: "Yes, delete it!"
            }).then(async (result) => {
                if (result.isConfirmed) {
                    Swal.fire({
                        timer: 1000,
                        timerProgressBar: true,
                        allowOutsideClick: false,
                        didOpen: () => {
                            Swal.showLoading();
                        }
                    });
                    await deleteDoc(notRef);
                    dispatch( deleteNote(id) );
                    Swal.fire({
                        title: "Deleted!",
                        text: "Your note has been deleted.",
                        icon: "success"
                    });
                }
              });
        } catch (e) {
            Swal.fire('Ha ocurrido un error durante la eliminación de la nota', e, 'error');
        }
         
    };
};

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: id
});

export const noteLogout = () => ({
    type: types.notesLogoutCleaning
})
