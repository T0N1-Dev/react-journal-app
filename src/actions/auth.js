import { types } from "../types/types";
import { createUserWithEmailAndPassword, signInWithEmailAndPassword, signInWithPopup, updateProfile } from "firebase/auth";
import { auth, googleAuthProvider } from "../firebase/firebaseConfig";
import { finishLoading, startLoading } from "./ui";
import Swal from 'sweetalert2';
import { noteLogout } from "./notes";

export const startLoginEmailPassword = (email, password) => {

    return async (dispatch) => {
        try {
            dispatch(startLoading());
            const {user} = await signInWithEmailAndPassword(auth, email, password);

            dispatch( login(user.uid, user.displayName));
            dispatch(finishLoading());
        }
        catch (err) {
            dispatch(finishLoading());
            Swal.fire('Error', err.message, 'error');
        } 
    }
}

export const startRegisterWithEmailPassswordName = (email, password, name) => {

    return async (dispatch) => {
        try {
            dispatch(startLoading());
            const { user } = await createUserWithEmailAndPassword(auth, email, password);
            await updateProfile(user, { displayName: name });
            
            dispatch(
                login(user.uid, user.displayName)
            );
            dispatch(finishLoading());
        }
        catch (err) {
            Swal.fire('Error', err.message, 'error');
            dispatch(finishLoading())
        } 
    }
}

export const startGoogleLogin = () => {
    return async (dispatch) => {
        try {
          const { user } = await signInWithPopup(auth, googleAuthProvider);
          dispatch(
            login( user.uid, user.displayName )
          );
        } catch (error) {
          console.error("Error al iniciar sesiÃ³n con Google:", error);
        }
    };
}

export const login = (uid, displayName) => ({
        type: types.login,
        payload: {
            uid,
            displayName
        }
    })

export const startLogout = () => {
    return async (dispatch) => {
        try {
            await auth.signOut();
            dispatch( logout() );
            dispatch( noteLogout() );
        } catch (err) {
            console.error("Error al hacer Logout:", err);
        }
    }
}

export const logout = () => ({
    type: types.logout
})