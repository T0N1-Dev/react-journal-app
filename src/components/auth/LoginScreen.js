import React from 'react';
import { Link } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { useDispatch, useSelector } from 'react-redux';
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';
import validator from 'validator';
import { setError } from '../../actions/ui';

export const LoginScreen = () => {
  const dispatch = useDispatch();
  const { mgError, loading } = useSelector((state) => state.ui);

  const [formValues, handleInputChange] = useForm({
    email: 'test@gmail.com',
    password: '123456',
  });

  const { email, password } = formValues;

  const handleLogin = (e) => {
    e.preventDefault();

    if (validator.isEmail(email)) {
      dispatch(startLoginEmailPassword(email, password));
    } else {
      dispatch(setError('Invalid email: ' + email));
    }
  };

  const handleGoogleLogin = () => {
    dispatch(startGoogleLogin());
  };

  return (
    <>
      <h3 className="auth__title">Login</h3>

      <form 
        onSubmit={handleLogin}
        className='animate__animated animate__fadeIn animate__faster'
      >
        {mgError && <div className="auth__alert-error">{mgError}</div>}

        <input
          type="text"
          placeholder="Email"
          name="email"
          className="auth__input"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />

        <input
          type="password"
          placeholder="Password"
          name="password"
          className="auth__input"
          value={password}
          onChange={handleInputChange}
        />

        <button
          type="submit"
          className="btn btn-primary btn-block"
          disabled={loading}
        >
          Login
        </button>

        <div className="auth__social-network">
          <p>Login with social networks</p>

          <div className="google-btn" onClick={handleGoogleLogin}>
            <div className="google-icon-wrapper">
              <img
                className="google-icon"
                src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQRB4kg1jgnqD5fEaCSHKMoUdFeTBj9CH_opA&s"
                alt="google button"
              />
            </div>

            <p className="btn-text">
              <b>Sign in with google</b>
            </p>
          </div>
        </div>

        <Link to="/auth/register" className="link" data-testid="create-link">
          Create new account
        </Link>
      </form>
    </>
  );
};

