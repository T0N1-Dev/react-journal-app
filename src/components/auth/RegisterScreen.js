import React from 'react'
import { Link } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import validator from 'validator';
import { useDispatch, useSelector } from 'react-redux';
import { removeError, setError } from '../../actions/ui';
import { startRegisterWithEmailPassswordName } from '../../actions/auth';

export const RegisterScreen = () => {

  const dispatch = useDispatch();
  const { mgError, loading } = useSelector( state => state.ui );

  const [ { name, email, password, password2 }, handleInputChange, reset ] = useForm({
    name: 'Prueba', 
    email:'prueba@gmail.com', 
    password:'123456', 
    password2:'123456'
  });

  const handleRegister = (e) => {
    e.preventDefault();

    if ( isFormValid() ){
      dispatch(startRegisterWithEmailPassswordName(email, password, name));
    }
    reset();
  };

  const isFormValid = () => {

    if ( name.trim().length === 0 || !validator.isAlpha(name) ) {
      dispatch(setError('Invalid name: ' + name));
      return false;
    } else if ( email.trim().length === 0 || !validator.isEmail(email) ) {
      dispatch(setError('Invalid email: ' + email));
      return false;
    } else if ( password !== password2 || password.length < 6) {
      dispatch(setError('Invalid passwords'));
      return false;
    }

    dispatch( removeError() );
    return true;
  }

  return (
    <>
      <h3 className='auth__title'>Register</h3>

      <form 
        onSubmit={ handleRegister }
        className='animate__animated animate__fadeIn animate__faster'
      >

      {
        mgError && 
        (
          <div className='auth__alert-error'>
            { mgError }
          </div>
        )
      }

      <input 
          type='text'
          placeholder='Name'
          name='name'
          className='auth__input'
          autoComplete='off'
          value={name}
          onChange={handleInputChange}
        />  

        <input 
          type='text'
          placeholder='Email'
          name='email'
          className='auth__input'
          autoComplete='off'
          value={email}
          onChange={handleInputChange}
        />  

        <input 
          type='password'
          placeholder='Password'
          name='password'
          className='auth__input'
          value={password}
          onChange={handleInputChange}
        />

        <input 
          type='password'
          placeholder='Confirm Password'
          name='password2'
          className='auth__input'
          value={password2}
          onChange={handleInputChange}
        />

        <button
          type='submit'
          className='btn btn-primary btn-block mb-5'
          disabled={ loading }
        >
          Register
        </button>

        <Link
        to="/auth/login"
        className="link"
        >
          Already registered?
        </Link>
      </form>  
    </>
  )
}
