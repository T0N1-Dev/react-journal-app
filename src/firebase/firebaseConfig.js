import { initializeApp } from "firebase/app";
import { connectFirestoreEmulator, getFirestore } from "firebase/firestore";
import { getAuth, GoogleAuthProvider } from "firebase/auth";

const firebaseConfig = {
  apiKey: process.env.REACT_APP_APIKEY,
  authDomain: process.env.REACT_APP_AUTHDOMAIN,
  projectId: process.env.REACT_APP_PROJECTID,
  storageBucket: process.env.REACT_APP_STORAGEBUCKET,
  messagingSenderId: process.env.REACT_APP_MESSAGINGSENDERID,
  appId: process.env.REACT_APP_APPID
};

const app = initializeApp(firebaseConfig);
const db = getFirestore(app);
const auth = getAuth(app);
const googleAuthProvider = new GoogleAuthProvider();


export { db, auth, googleAuthProvider };