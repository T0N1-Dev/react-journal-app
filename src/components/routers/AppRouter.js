import React, { useEffect, useState } from 'react';
import { BrowserRouter, Switch, Redirect } from 'react-router-dom';
import { JournalScreen } from '../journal/JournalScreen';
import { AuthRouter } from './AuthRouter';
import { auth } from '../../firebase/firebaseConfig';
import { useDispatch } from 'react-redux';
import { login } from '../../actions/auth';
import { PrivateRoute } from './PrivateRoutes';
import { PublicRoute } from './PublicRoutes';
import { startLoadingNotes } from '../../actions/notes';

export const AppRouter = () => {
  const dispatch = useDispatch();
  const [ isLoggedIn, setIsLoggedIn ] = useState(false);
  const [checking, setChecking] = useState(true);

  useEffect(() => {
    auth.onAuthStateChanged(async (user) => {
      if (user?.uid) {
        dispatch(login(user.uid, user.displayName));
        setIsLoggedIn(true);

        dispatch(startLoadingNotes(user.uid));
        
      } else {
        setIsLoggedIn(false);
      }
      setChecking(false);
    });
  }, [dispatch, setChecking]);

  if (checking) {
    return <h1>Wait...</h1>;
  }

  return (
      <BrowserRouter>
        <Switch>
          <PublicRoute 
            path="/auth/*" 
            component={AuthRouter}
            isAuthenticated={ isLoggedIn }
          />
          <PrivateRoute 
            exact 
            path="/" 
            component={JournalScreen}
            isAuthenticated={isLoggedIn} 
          />
          <Redirect to="/auth/login" />
        </Switch>
      </BrowserRouter>
  );
};

