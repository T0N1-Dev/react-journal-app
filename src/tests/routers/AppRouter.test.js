import { configureStore } from "@reduxjs/toolkit";
import { login } from "../../actions/auth";
import { authReducer } from "../../reducers/authReducer";
import { notesReducer } from "../../reducers/notesReducer";
import { uiReducer } from "../../reducers/uiReducer";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom/cjs/react-router-dom.min";
import { AppRouter } from "../../components/routers/AppRouter";
import { act } from 'react';
import { signInWithEmailAndPassword } from "firebase/auth";
import { auth } from "../../firebase/firebaseConfig";

jest.mock('../../actions/auth', () => ({
    login: jest.fn()
}));

describe('Pruebas en <AppRouter />', () => {

    let store;
    let contenedor;

    beforeEach(() => {
        store = configureStore({
            reducer: {
                auth: authReducer,
                notes: notesReducer,
                ui: uiReducer
            },
            preloadedState: {
                auth: {},
                notes: { notes: [], active: null},
                ui: {}
            }
        });

        store.dispatch = jest.fn();

        const { container } = render(
            <Provider store={store}>
                <MemoryRouter>
                    <AppRouter />
                </MemoryRouter>
            </Provider>
        );

        contenedor = container;

        jest.clearAllMocks();
    });
    
    test('should match with snapshot', () => { 
        expect(contenedor).toMatchSnapshot();
    })

    test('should trigger the login action', async () => {
        
        let user;

        await act( async () => {
            const userCred = await signInWithEmailAndPassword(auth, 'test.testing@gmail.com', '123456');
            user = userCred.user;   
        });

        render(
            <Provider store={store}>
                <MemoryRouter>
                    <AppRouter />
                </MemoryRouter>
            </Provider>
        );

        expect(login).toHaveBeenCalledWith(user.uid, user.displayName);
    });
});