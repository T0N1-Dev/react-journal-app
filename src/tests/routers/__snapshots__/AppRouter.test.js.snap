// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AppRouter /> should match with snapshot 1`] = `
<div>
  <h1>
    Wait...
  </h1>
</div>
`;
